File: 10-Recursion-and-Backtracking/python/02_n_queens.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        result = []
        queens = [0] * n

        def backtrack(row, cols, diags, anti_diags):
            if row == n:
                board = []
                for r in range(n):
                    row_str = ['.'] * n
                    row_str[queens[r]] = 'Q'
                    board.append("".join(row_str))
                result.append(board)
                return

            available = ((1 << n) - 1) & ~(cols | diags | anti_diags)
            while available:
                position = available & (-available)
                col = (position - 1).bit_length()
                queens[row] = col

                backtrack(
                    row + 1,
                    cols | position,
                    (diags | position) << 1,
                    (anti_diags | position) >> 1
                )

                available &= available - 1

        backtrack(0, 0, 0, 0)
        return result

File: 10-Recursion-and-Backtracking/cpp/02-n-queens.cpp
class Solution {
public:
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> result;
        vector<int> queens(n);
        backtrack(0, n, 0, 0, 0, queens, result);
        return result;
    }

private:
    void backtrack(int row, int n, int cols, int diags, int antiDiags,
                   vector<int>& queens, vector<vector<string>>& result) {
        if (row == n) {
            vector<string> board(n, string(n, '.'));
            for (int i = 0; i < n; i++) {
                board[i][queens[i]] = 'Q';
            }
            result.push_back(board);
            return;
        }

        int availablePositions = ((1 << n) - 1) & ~(cols | diags | antiDiags);
        while (availablePositions != 0) {
            int position = availablePositions & (-availablePositions);
            int col = __builtin_ctz(position);
            queens[row] = col;

            backtrack(row + 1, n, cols | position,
                     (diags | position) << 1,
                     (antiDiags | position) >> 1,
                     queens, result);

            availablePositions &= availablePositions - 1;
        }
    }
};
